cdap-securestore-ext-gcp-secretstore/src/main/java/io/cdap/cdap/securestore/gcp/cloudsecretmanager/GcpSecretManager.java

```diff
@@ -0,0 +1,111 @@
+import com.google.api.gax.rpc.StatusCode;
+import io.cdap.cdap.securestore.spi.SecretManager;
+import io.cdap.cdap.securestore.spi.SecretManagerContext;
+import io.cdap.cdap.securestore.spi.SecretNotFoundException;
+import io.cdap.cdap.securestore.spi.secret.Secret;
+import io.cdap.cdap.securestore.spi.secret.SecretMetadata;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.stream.Collectors;
+
+/**
+ * SecretManager implementation backed by GCP's service also called "Secret Manager"
+ * https://cloud.google.com/secret-manager
+ *
+ * <p>This implementation doesn't store any infromation within the CDAP database and instead stores
+ * the credentials entirely inside GCP Secret Manager -- this allows credentials to be shared
+ * between multiple CDAP instances without additional coordination.
+ *
+ * <p>Due to Secret Manager annotation <a
+ * href="https://cloud.google.com/secret-manager/docs/reference/rest/v1/projects.secrets#Secret">limitations</a>
+ * the total size of metadata associated with any given secret cannot exceed 16 KiB.
+ */
+public class GcpSecretManager implements SecretManager {
+  private static final String PROVIDER_NAME = "gcp-secretmanager";
+
+  private CloudSecretManagerClient client;
+
+  @Override
+  public String getName() {
+    return PROVIDER_NAME;
+  }
+
+  @Override
+  public void initialize(SecretManagerContext context) throws IOException {
+    this.client = new CloudSecretManagerClient(context.getProperties());
+  }
+
+  @Override
+  public void store(String namespace, Secret secret) throws IOException {
+    WrappedSecret wrappedSecret = WrappedSecret.fromMetadata(namespace, secret.getMetadata());
+
+    try {
+      Secret existingSecret = get(namespace, secret.getMetadata().getName());
+      // 'update' only if 'get' request succeeds, otherwise 'create'.
+      client.updateSecret(wrappedSecret);
+
+      // Add a new secret version only if the secret payload has changed.
+      if (!Arrays.equals(existingSecret.getData(), secret.getData())) {
+        client.addSecretVersion(wrappedSecret, secret.getData());
+      }
+    } catch (SecretNotFoundException e) {
+      client.createSecret(wrappedSecret);
+      client.addSecretVersion(wrappedSecret, secret.getData());
+    } catch (ApiException e) {
+      throw new IOException("Secret Manager API Call failed", e);
+    }
+  }
+
+  @Override
+  public Secret get(String namespace, String name) throws SecretNotFoundException, IOException {
+    try {
+      WrappedSecret wrappedSecret = client.getSecret(namespace, name);
+      byte[] secretData = client.getSecretData(wrappedSecret);
+      return new Secret(secretData, wrappedSecret.getCdapSecretMetadata());
+    } catch (ApiException e) {
+      if (e.getStatusCode().getCode() == StatusCode.Code.NOT_FOUND) {
+        throw new SecretNotFoundException(namespace, name);
+      }
+      throw new IOException("Secret Manager API call failed", e);
```

Comment:
nit: it's a little better to include what type of call in the message, like 'Secret Manager Get call failed'

Reply:
done.